### PR TITLE
uint256 division

### PR DESCRIPTION
**Pull request by spapinistarkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)

### CODE REVIEW

**Comment by spapinistarkware** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2756** <a href="https://app.graphite.dev/github/pr/starkware-libs/cairo/2756" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/starkware-libs/cairo/2756?utm_source=stack-comment).

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 3 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 3 of 15 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

___
*[`corelib/src/integer.cairo` line 993 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSe4J5w8s9hi5OpjwRu:-NSe4J5x16blzBehB-i6:bh2xvp8) ([raw file](https://github.com/starkware-libs/cairo/blob/6c6471b713cc4acf827309932f74e978e98021a8/corelib/src/integer.cairo#L993)):*
> ```Cairo
> 
> impl U256Div of Div<u256> {
>     fn div(a: u256, b: u256) -> u256 {
> ```

add Rem as well.



_Code quote:_
```Cairo
impl U256Div of Div<u256> {
    fn div(a: u256, b: u256) -> u256 {
        let (q, r) = u256_safe_divmod(a, u256_as_non_zero(b));
        q
    }
}
impl U256DivEq of DivEq<u256> {
    #[inline(always)]
    fn div_eq(ref self: u256, other: u256) {
        self = Div::div(self, other);
    }
}
```
___
*[`crates/cairo-lang-casm/src/hints/mod.rs` line 35 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSe4cHFC-Mum9jeYojY:-NSe4cHFC-Mum9jeYojZ:b1tkzx0) ([raw file](https://github.com/starkware-libs/cairo/blob/6c6471b713cc4acf827309932f74e978e98021a8/crates/cairo-lang-casm/src/hints/mod.rs#L35)):*
> ```XML
>     Uint256DivMod {
>         dividend_low: ResOperand,
>         dividend_high: ResOperand,
> ```

doc - as this is non trivial (especially the `extra`s)



_Code quote:_
```XML
    Uint256DivMod {
        dividend_low: ResOperand,
        dividend_high: ResOperand,
        divisor_low: ResOperand,
        divisor_high: ResOperand,
        quotient0: CellRef,
        quotient1: CellRef,
        divisor0: CellRef,
        divisor1: CellRef,
        extra0: CellRef,
        extra1: CellRef,
        remainder_low: CellRef,
        remainder_high: CellRef,
    },
```

**Review by orizi — COMMENTED** [review]
> Reviewed 3 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 6 of 15 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

**Review by orizi — COMMENTED** [review]
> Reviewed 2 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 8 of 15 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

**Review by orizi — COMMENTED** [review]
> Reviewed 1 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 9 of 15 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 9 of 15 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @orizi)

___
*[`corelib/src/integer.cairo` line 993 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSe4J5w8s9hi5OpjwRu:-NSe5tgY0vchLGEi1dKM:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/6c6471b713cc4acf827309932f74e978e98021a8/corelib/src/integer.cairo#L993)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

add Rem as well.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-casm/src/hints/mod.rs` line 35 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSe4cHFC-Mum9jeYojY:-NSe7VXf2U_-gqmYk1qQ:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/6c6471b713cc4acf827309932f74e978e98021a8/crates/cairo-lang-casm/src/hints/mod.rs#L35)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

doc - as this is non trivial (especially the `extra`s)
</blockquote></details>

Done.

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 9 of 16 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

___
*[`crates/cairo-lang-casm/src/hints/mod.rs` line 36 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeKw3n8UALQhsghbis:-NSeKw3n8UALQhsghbit:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-casm/src/hints/mod.rs#L36)):*
> ```XML
>     /// Splits the remainder to 128bit words: remainder_low and remainder_high.
>     /// Splits the quotient and the divisor to 128bit words.
>     /// The lower 128 bits of the quotient are written to quotient0 and quotient1.
> ```





_Suggestion:_
```XML
    /// Divides dividend_low<<128+dividend_high by divisor_low<<128+divisor_high.
    /// Splits the remainder to 128bit words: remainder_low and remainder_high.
    /// Splits the quotient and the divisor to 128bit words.
    /// The lower 128 bits of the quotient are written to quotient0 and quotient1.
    /// The lower 128 bits of the divisor are written to divisor0 and divisor1.
    /// If the divisor is greater than 2^128, the upper 128 bits of the divisor are written to
    /// extra0 and extra1. In this case, quotient must be lower than 2^128.
    /// Otherwise, the upper 128 bits of the quotient are written to extra0 and extra1.
```
___
*[`crates/cairo-lang-runner/src/casm_run/mod.rs` line 267 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeLYaMEGRvrDiZ4kUo:-NSeLYaN3grVNtRJAB5_:b-y4djsc) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-runner/src/casm_run/mod.rs#L267)):*
> ```XML
>                 // Guess quotient limbs.
>                 let (limb, quotient) =
>                     (quotient.clone() % pow_2_64.clone(), quotient / pow_2_64.clone());
> ```

and everywhere.



_Suggestion:_
```XML
                let (limb, quotient) = quotient.div_rem(quotient, pow_2_64.clone());
```

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 10 of 16 files reviewed, 5 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSeOr982p_AYooDKm8Y:b-tinjxi) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
> ```XML
> 
>         // Check consistency.
>         tempvar shifted_divisor1 = divisor1 * u64_limit;
> ```

maybe we should use the simple divmod hint for these.



_Code quote:_
```XML
        // Check consistency.
        tempvar shifted_divisor1 = divisor1 * u64_limit;
        assert divisor_low = divisor0 + shifted_divisor1;

        // Push quotient_low for output.
        tempvar shifted_quotient1 = quotient1 * u64_limit;
        assert quotient_low = quotient0 + shifted_quotient1;
```
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 157 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeQVTV4Jp-gBsAXV5a:-NSeQVTV4Jp-gBsAXV5b:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L157)):*
> ```XML
> 
>     casm_build_extend! {casm_builder,
>         jump DivisorIsSmall if divisor_high != 0;
> ```





_Suggestion:_
```XML
DivisorIsLarge
```
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 158 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeQT0L-q0QgMGv7H0v:-NSeQT0M-QAEh5qI4ZFE:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L158)):*
> ```XML
>     casm_build_extend! {casm_builder,
>         jump DivisorIsSmall if divisor_high != 0;
> ```





_Suggestion:_
```XML
// Here, quotient is large.
```

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 10 of 16 files reviewed, 5 unresolved discussions (waiting on @liorgold2 and @orizi)

___
*[`crates/cairo-lang-casm/src/hints/mod.rs` line 36 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeKw3n8UALQhsghbis:-NSlA7uVAUSw-lUpKtiI:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-casm/src/hints/mod.rs#L36)):*
> ```XML
>     /// Splits the remainder to 128bit words: remainder_low and remainder_high.
>     /// Splits the quotient and the divisor to 128bit words.
>     /// The lower 128 bits of the quotient are written to quotient0 and quotient1.
> ```

Done.
___
*[`crates/cairo-lang-runner/src/casm_run/mod.rs` line 267 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeLYaMEGRvrDiZ4kUo:-NSlAmBG6abEv3AdBzsw:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-runner/src/casm_run/mod.rs#L267)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

and everywhere.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSlB29k0Oe6LCtq0p_V:b9y7aya) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

maybe we should use the simple divmod hint for these.
</blockquote></details>

Don't know what to answer :S
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 157 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeQVTV4Jp-gBsAXV5a:-NSlAkr52W_5zZCP_GGj:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L157)):*
> ```XML
> 
>     casm_build_extend! {casm_builder,
>         jump DivisorIsSmall if divisor_high != 0;
> ```

Done.
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 158 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeQT0L-q0QgMGv7H0v:-NSlAlEk6B_oMMqyTVby:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L158)):*
> ```XML
>     casm_build_extend! {casm_builder,
>         jump DivisorIsSmall if divisor_high != 0;
> ```

Done.

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 15 files at r1, 2 of 4 files at r2, 10 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 17 of 18 files reviewed, 3 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

___
*[`crates/cairo-lang-sierra/src/extensions/modules/uint256.rs` line 85 at r3](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSmJOsX2UuEBs7eynG7:-NSmJOsX2UuEBs7eynG8:b-aek0ou) ([raw file](https://github.com/starkware-libs/cairo/blob/eec2580c381fa5fb735fbd7427e5f22c9a3d1466/crates/cairo-lang-sierra/src/extensions/modules/uint256.rs#L85)):*
> ```XML
>                 OutputVarInfo {
>                     ty: ty.clone(),
>                     ref_info: OutputVarReferenceInfo::Deferred(DeferredOutputKind::Generic),
> ```

as you write these to ap.



_Suggestion:_
```XML
                OutputVarInfo {
                    ty: ty.clone(),
                    ref_info: OutputVarReferenceInfo::NewTemp,
                },
                OutputVarInfo {
                    ty,
                    ref_info: OutputVarReferenceInfo::NewTemp,
```
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSmJkoh488xI71mBxVS:bo1rkp6) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

Don't know what to answer :S
</blockquote></details>

let's find another opinion @liorgold2

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 17 of 18 files reviewed, 3 unresolved discussions (waiting on @liorgold2 and @orizi)

___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSmKa8GE992bnnVDeOK:bhzjnhj) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

let's find another opinion @liorgold2 
</blockquote></details>

I didn't understand your comment :( Can you elaborate?

**Review by orizi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 17 of 18 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSmUwq59LD4e2H7y6sz:bo60860) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

I didn't understand your comment :( Can you elaborate?
</blockquote></details>

you hint splits into 64 bits things that other hints can already do (the basic divmod) - i'm wondering if at least these parts should be done in another hint.

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 17 of 18 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @orizi)

___
*[`crates/cairo-lang-sierra/src/extensions/modules/uint256.rs` line 85 at r3](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSmJOsX2UuEBs7eynG7:-NSqoSSGDipj6TFxV8Ip:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/eec2580c381fa5fb735fbd7427e5f22c9a3d1466/crates/cairo-lang-sierra/src/extensions/modules/uint256.rs#L85)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

as you write these to ap.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSqodY-AYzJJ1sTTAz6:b1i2hvv) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

you hint splits into 64 bits things that other hints can already do (the basic divmod) - i'm wondering if at least these parts should be done in another hint. 
</blockquote></details>

This is more efficinet though, right? Less traveling back and forth. Also, I want to merge this PR already. It's really needed...

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 17 of 18 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @spapinistarkware)

___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NSrRbnx2u6JmnwdfsN1:bubqsua) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

This is more efficinet though, right? Less traveling back and forth. Also, I want to merge this PR already. It's really needed...
</blockquote></details>

This might be a little more efficient (but not in the actual casm) - but this code is still less readable than it could probably be.
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 132 at r4](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSrS-C21xJ0uTlo2xF9:-NSrS-C38-6FDQTPFmDf:biwowm) ([raw file](https://github.com/starkware-libs/cairo/blob/c51d4b23fa0842af418a1d7f901965d6aab89f00/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L132)):*
> ```XML
> 
>         // Multiply in parts.
>         // Limb0.
> ```

Expand names or add doc at the limb calculation as it is rather hard to follow

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: 17 of 18 files reviewed, 2 unresolved discussions (waiting on @liorgold2 and @orizi)

___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NStJknB8_3VmpV16anc:b-h2sts6) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

This might be a little more efficient (but not in the actual casm) - but this code is still less readable than it could probably be.
</blockquote></details>

I tried it and it's a lot less clear IMO. There are multiple hint calls and extra0,extra1 still need to be in the original hint. And the divmod hints just divis by 64, it's a lot more efficient to do shift.
___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 132 at r4](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSrS-C21xJ0uTlo2xF9:-NStGGtNApTQDmZ2Bp6I:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/c51d4b23fa0842af418a1d7f901965d6aab89f00/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L132)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

Expand names or add doc at the limb calculation as it is rather hard to follow
</blockquote></details>

Done.

**Review by orizi — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756#-:-NStMK671pE_cl0RAn6U:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @liorgold2)

___
*[`crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs` line 124 at r2](https://reviewable.io/reviews/starkware-libs/cairo/2756#-NSeOr982p_AYooDKm8X:-NStM0SY6WGTz7yr6PNy:b32wtd) ([raw file](https://github.com/starkware-libs/cairo/blob/995be3160f12ae97ab9468423318ab7f913780ae/crates/cairo-lang-sierra-to-casm/src/invocations/uint256.rs#L124)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

I tried it and it's a lot less clear IMO. There are multiple hint calls and extra0,extra1 still need to be in the original hint. And the divmod hints just divis by 64, it's a lot more efficient to do shift.
</blockquote></details>

ok

**Review by orizi — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2756)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @liorgold2)
